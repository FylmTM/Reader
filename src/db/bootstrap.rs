use rusqlite::{Connection, NO_PARAMS};

const SCHEMA_VERSION: u16 = 1;
// language=SQLite
const SCHEMA_VERSION_QUERY: &str = "
create table if not exists schema_version
(
    id      integer primary key,
    version integer
);

insert or ignore into schema_version(id, version)
values (1, 0);
";

pub fn initialize_schema(conn: &Connection) {
    conn.execute_batch(SCHEMA_VERSION_QUERY)
        .expect("Failed to create schema_version table.");

    let mut current_version: u16 = conn
        .query_row(
            // language=SQLite
            "select version from schema_version",
            NO_PARAMS,
            |row| row.get(0),
        )
        .expect("Failed to retrieve current schema version.");

    let version_before = current_version;

    info!(
        "Current schema version=V{}, required=V{}.",
        current_version, SCHEMA_VERSION
    );

    if current_version == 0 {
        info!("Applying V1 migration.");
        conn.execute_batch(include_str!("migrations/V1_initial.sql"))
            .expect("Schema creation failed.");
        current_version = 1;
    }

    if version_before != current_version {
        info!("Set current schema version to V{}.", current_version);
        conn.execute(
            // language=SQLite
            "update schema_version set version = ? where id = 1",
            &[current_version],
        )
        .expect("Failed to update current schema version.");
    }
}

pub fn initialize_fixture(conn: &Connection) {
    // language=SQLite
    let query = "
        insert or ignore into users(id, username, api_key)
        values (1, 'me', 'api_key');
    ";
    conn.execute_batch(query).expect("Failed to apply fixture.");
}
