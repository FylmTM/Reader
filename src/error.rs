use std::io::Cursor;

use rocket::http::Status;
use rocket::response::Responder;
use rocket::{Request, Response};

use ApplicationError::*;

use crate::api::ApiError;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug)]
pub enum Error {
    App(ApplicationError),
    Pool(r2d2::Error),
    DB(rusqlite::Error),
    HttpClient(reqwest::Error),
    RSS(rss::Error),
    Atom(atom_syndication::Error),
    Chrono(chrono::ParseError),
}

#[derive(Debug)]
pub enum ApplicationError {
    AuthenticationApiKeyNotFound,
    FailedGetUrlContent(String),
    RSSParsingMissingItemLink,
    RSSParsingMissingItemTitle,
    AtomParsingMissingEntryLink,
    QueryEntityAlreadyExists,
    QueryEntityNotFound,
    QueryUnexpectedRowCount { expected: i64, actual: i64 },
}

impl Error {
    fn describe(self) -> (Status, String) {
        match self {
            Error::App(AuthenticationApiKeyNotFound) => {
                (Status::NotFound, "API key does not exists.".to_string())
            }
            Error::App(QueryUnexpectedRowCount { expected, actual }) => (
                Status::BadRequest,
                format!("Expected row count {}, got {}.", expected, actual),
            ),
            Error::DB(rusqlite::Error::QueryReturnedNoRows) => {
                (Status::NotFound, "Entity not found.".to_string())
            }
            _ => (Status::InternalServerError, format!("{:?}", self)),
        }
    }
}

impl<'r> Responder<'r> for Error {
    fn respond_to(self, _: &Request) -> std::result::Result<Response<'r>, Status> {
        let (http_status, message) = self.describe();

        let json = serde_json::to_string(&ApiError::with(message))
            .expect("Failed to serialize api error.");

        Response::build()
            .header(rocket::http::ContentType::JSON)
            .sized_body(Cursor::new(json))
            .status(http_status)
            .ok()
    }
}

impl From<rusqlite::Error> for Error {
    fn from(error: rusqlite::Error) -> Error {
        Error::DB(error)
    }
}

impl From<r2d2::Error> for Error {
    fn from(error: r2d2::Error) -> Error {
        Error::Pool(error)
    }
}

impl From<reqwest::Error> for Error {
    fn from(error: reqwest::Error) -> Error {
        Error::HttpClient(error)
    }
}

impl From<rss::Error> for Error {
    fn from(error: rss::Error) -> Error {
        Error::RSS(error)
    }
}

impl From<atom_syndication::Error> for Error {
    fn from(error: atom_syndication::Error) -> Error {
        Error::Atom(error)
    }
}

impl From<chrono::ParseError> for Error {
    fn from(error: chrono::ParseError) -> Error {
        Error::Chrono(error)
    }
}
