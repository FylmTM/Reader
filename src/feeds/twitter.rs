use rusqlite::Connection;
use std::thread;
use std::time::Duration;

pub fn update(_conn: &Connection) {
    thread::sleep(Duration::from_secs(5));
}
