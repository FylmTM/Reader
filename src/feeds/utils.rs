lazy_static! {
    static ref AMMONIA_TEXT: ammonia::Builder<'static> = {
        let mut builder = ammonia::Builder::default();
        builder.tags(std::collections::HashSet::new());
        builder
    };
}

pub fn clean_to_text(html: &str) -> String {
    AMMONIA_TEXT.clean(html).to_string()
}

pub fn clean_to_safe_html(html: &str, base_url: &str) -> String {
    let mut builder = ammonia::Builder::default();
    builder.url_relative(ammonia::UrlRelative::RewriteWithBase(
        ammonia::Url::parse(base_url).expect("Failed to parse base url"),
    ));
    builder.clean(html).to_string()
}

pub fn find_image(html: &str) -> Option<String> {
    // language=RegExp
    let regexp = regex::Regex::new(r#"<img.*?src="(.+?)""#).unwrap();
    regexp
        .captures(html)
        .and_then(|captures| captures.get(1))
        .map(|src| src.as_str().to_string())
}
