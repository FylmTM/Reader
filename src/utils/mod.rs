use rocket::fairing::{Fairing, Info, Kind};
use rocket::{Request, Response};

pub struct AppHeaders();

impl Fairing for AppHeaders {
    fn info(&self) -> Info {
        Info {
            name: "Application Headers",
            kind: Kind::Response,
        }
    }

    fn on_response(&self, _request: &Request, response: &mut Response) {
        response.set_raw_header(
            "Content-Security-Policy",
            "default-src 'none'; script-src 'self' 'unsafe-inline'; connect-src 'self'; img-src *; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com",
        );
    }
}
