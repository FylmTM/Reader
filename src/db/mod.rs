use rusqlite::{Connection, NO_PARAMS};

pub use self::entities::Post;
use crate::error::Error;

mod entities;

pub fn initialize(conn: &Connection) {
    conn.execute(include_str!("queries/schema.sql"), NO_PARAMS)
        .expect("schema created");
}

pub fn find_posts(conn: &Connection) -> Result<Vec<Post>, Error> {
    let mut statement = conn.prepare("SELECT id, title, body FROM posts")?;
    let result = statement.query_map(NO_PARAMS, |row| {
        Ok(Post {
            id: row.get(0)?,
            title: row.get(1)?,
            body: row.get(2)?,
        })
    })?;

    let mut posts = Vec::new();
    for post in result {
        posts.push(post?);
    }
    Ok(posts)
}
