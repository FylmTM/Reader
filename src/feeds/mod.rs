use std::thread;
use std::time::Duration;

use crate::db::{self, Queries};
use crate::error::{ApplicationError, Error, Result};

pub mod atom;
pub mod rss;
pub mod utils;

pub fn initialize(pool: &db::Pool, interval: u64) {
    let pool = pool.clone();

    info!("Schedule feeds update every {} minutes.", interval);
    thread::spawn(move || loop {
        thread::sleep(Duration::from_secs(5));
        loop {
            let mut conn = pool.get().expect("Failed to acquire db connection");
            update(&mut conn);

            info!("Next feed update in {} minutes.", interval);
            thread::sleep(Duration::from_secs(60 * interval));
        }
    });
}

pub fn update(mut conn: &mut db::Connection) {
    info!("Feeds update started.");
    match conn.find_feeds() {
        Ok(feeds) => {
            for feed in feeds {
                match update_feed(&mut conn, &feed) {
                    Ok(_) => debug!("{:?} feed updated.", feed),
                    Err(error) => error!("{:?} feed failed to update: {:?}", feed, error),
                };
            }
        }
        Err(error) => {
            error!("Failed to retrieve feeds: {:?}.", error);
        }
    }
    info!("Feeds update completed.");
}

pub fn update_feed(conn: &mut db::Connection, feed: &db::Feed) -> Result<()> {
    debug!("{:?} update started.", feed);

    let user_ids = conn.find_user_ids_by_feed(feed.id)?;
    if user_ids.is_empty() {
        debug!("{:?} does not belong to any user, skip update.", feed);
        return Ok(());
    }

    let posts = match feed.kind {
        db::FeedKind::RSS => rss::update(&feed),
        db::FeedKind::Atom => atom::update(&feed),
    }?;

    save_posts(conn, feed, &posts, &user_ids)?;

    Ok(())
}

fn save_posts(
    conn: &mut db::Connection,
    feed: &db::Feed,
    posts: &Vec<db::Post>,
    user_ids: &Vec<db::UserId>,
) -> Result<()> {
    for post in posts {
        let tx = conn.transaction()?;
        match tx.save_post(feed.id, &post) {
            Ok(post_id) => {
                debug!("{:?} saved.", post);
                for user_id in user_ids {
                    tx.save_user_post(*user_id, post_id)?;
                }
                tx.commit()?;
            }
            Err(Error::App(ApplicationError::QueryEntityAlreadyExists)) => {
                debug!("{:?} already exists.", &post)
            }
            Err(error) => return Err(error),
        }
    }
    Ok(())
}
