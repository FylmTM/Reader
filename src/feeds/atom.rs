use crate::db::{Feed, MediaType, Post};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils::{clean_to_safe_html, clean_to_text, find_image, get_url_content};

pub fn update(feed: &Feed) -> Result<Vec<Post>> {
    let content = get_url_content(&feed.feed)?;
    let posts = parse(&content)?;
    Ok(posts)
}

fn parse(content: &str) -> Result<Vec<Post>> {
    let feed = content.parse::<atom_syndication::Feed>()?;
    let mut posts = Vec::with_capacity(feed.entries().len());

    for entry in feed.entries() {
        let unique_id = entry.id().to_string();
        let link = entry
            .links()
            .iter()
            .find(|link| link.rel() == "alternate")
            .or_else(|| entry.links().get(0))
            .map(|link| link.href())
            .ok_or(Error::App(ApplicationError::AtomParsingMissingEntryLink))?
            .to_string();
        let title = entry.title().to_string();
        let date = parse_date(entry.published().unwrap_or(entry.updated()))?;
        let summary = entry.summary().map(clean_to_text);
        let content = entry
            .content()
            .and_then(|content| content.value())
            .or(entry.summary())
            .map(|content| clean_to_safe_html(content, &link));
        let comments_link = None;

        let (media_type, media_link) = {
            content
                .as_ref()
                .and_then(|content| find_image(content))
                .map(|image| {
                    (
                        Some(MediaType {
                            mime: "image/*".to_string(),
                        }),
                        Some(image),
                    )
                })
                .unwrap_or((None, None))
        };

        let post = Post {
            id: -1, // Post is not retrieved from database
            unique_id,
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();

    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result =
        chrono::DateTime::parse_from_rfc3339(date).map(|date| chrono::DateTime::from(date))?;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::atom::parse;

    #[test]
    fn test_parse_atom_github_rocket_releases() {
        let content = include_str!("../../static/mock/atom/github_rocket_releases.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_golang() {
        let content = include_str!("../../static/mock/atom/golang.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_julia_evans() {
        let content = include_str!("../../static/mock/atom/julia_evans.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_rust() {
        let content = include_str!("../../static/mock/atom/rust.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }
}
