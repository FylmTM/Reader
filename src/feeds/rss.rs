use crate::db::{self, Queries};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils;

pub fn update(conn: &mut db::Connection, feed: &db::Feed) -> Result<()> {
    let user_ids = conn.find_user_ids_by_feed(feed.id)?;
    if user_ids.is_empty() {
        debug!("{:?} does not belong to any user, skip update.", feed);
        return Ok(());
    }
    let content = get(&feed.feed)?;
    let posts = parse(&content)?;

    for post in posts {
        let tx = conn.transaction()?;

        match tx.save_post(feed.id, &post) {
            Ok(post_id) => {
                debug!("{:?} saved.", post);
                for user_id in &user_ids {
                    tx.save_user_post(*user_id, post_id)?;
                }
                tx.commit()?;
            }
            Err(Error::App(ApplicationError::QueryEntityAlreadyExists)) => {
                debug!("{:?} already exists.", &post)
            }
            Err(error) => return Err(error),
        }
    }

    Ok(())
}

fn get(link: &String) -> Result<String> {
    let mut response = reqwest::get(link)?;

    if response.status() != reqwest::StatusCode::OK {
        error!(
            "Failed to retrieve feed={} with status={}",
            link,
            response.status()
        );
        return Err(Error::App(ApplicationError::RSSFailedToRetrieve(
            link.clone(),
        )));
    }

    return Ok(response.text()?);
}

fn parse(content: &String) -> Result<Vec<db::Post>> {
    let channel = rss::Channel::read_from(content.as_bytes())?;

    let mut posts = Vec::with_capacity(channel.items().len());

    for item in channel.items() {
        let link = item
            .link()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemLink))?
            .to_string();
        let title = item
            .title()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemLink))?
            .to_string();
        let date = match item.pub_date() {
            Some(date) => parse_date(date)?,
            None => chrono::DateTime::from(chrono::Utc::now()),
        };

        let summary = item.description().map(utils::clean_to_text);
        let content = item
            .content()
            .or(item.description())
            .map(|content| utils::clean_to_safe_html(content, &link));
        let comments_link = item.comments().map(String::from);

        let media_type = item.enclosure().and_then(|enclosure| {
            let mime_type = enclosure.mime_type();
            if mime_type.is_empty() {
                return None;
            }
            if mime_type.parse::<mime::Mime>().is_err() {
                return None;
            }
            Some(db::MediaType {
                mime: mime_type.to_string(),
            })
        });
        let media_link = item.enclosure().and_then(|enclosure| {
            let url = enclosure.url();
            if url.is_empty() {
                None
            } else {
                Some(url.to_string())
            }
        });

        let post = db::Post {
            id: -1, // Post is not retrieved from database
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();
    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result = chrono::DateTime::parse_from_rfc2822(date)
        .or_else(|_| {
            // chrono consider "-0000" timezone to be "missing"
            // We will blindly replace it with +0000 if it exists, and try to parse again.
            let date = date.replace("-0000", "+0000");
            chrono::DateTime::parse_from_rfc2822(&date)
        })
        .map(|date| chrono::DateTime::from(date))?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::rss::{parse, parse_date};

    #[test]
    fn test_parse_rss_hacker_news() {
        let content = include_str!("../../static/mock/rss/hacker_news.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_kotlin() {
        let content = include_str!("../../static/mock/rss/kotlin.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_react() {
        let content = include_str!("../../static/mock/rss/react.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_stopgame() {
        let content = include_str!("../../static/mock/rss/stopgame.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_xkcd() {
        let content = include_str!("../../static/mock/rss/xkcd.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_date_handle_negative_zero_timezone() {
        let expected = chrono::DateTime::parse_from_rfc3339("2019-12-04T05:00:00+00:00").unwrap();
        assert_eq!(
            expected,
            parse_date("Wed, 04 Dec 2019 05:00:00 -0000").unwrap()
        );
    }
}
