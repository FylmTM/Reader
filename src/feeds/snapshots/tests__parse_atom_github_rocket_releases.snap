---
source: src/feeds/atom.rs
expression: result
---
[
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.3.14",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.14",
        title: "Rocket v0.3.14",
        date: 2018-06-23T06:39:06Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.3.14/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.3.15",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.15",
        title: "Rocket v0.3.15",
        date: 2018-07-16T08:23:30Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.3.15/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.3.16",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.16",
        title: "Rocket v0.3.16",
        date: 2018-08-24T07:33:16Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.3.16/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.3.17",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.17",
        title: "Rocket v0.3.17",
        date: 2018-09-30T01:05:28Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.3.17/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.4.0-rc.1",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0-rc.1",
        title: "Rocket v0.4-rc.1",
        date: 2018-10-31T22:09:18Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.4.0-rc.1/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> and <a href=\"https://rocket.rs/v0.4/news/2018-10-31-version-0.4-rc/\" rel=\"noopener noreferrer\">news article</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.4.0-rc.2",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0-rc.2",
        title: "Rocket v0.4-rc.2",
        date: 2018-12-01T05:25:08Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.4.0-rc.2/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> and <a href=\"https://rocket.rs/v0.4/news/2018-11-30-version-0.4-rc-2/\" rel=\"noopener noreferrer\">news article</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.4.0",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0",
        title: "Rocket v0.4.0",
        date: 2018-12-08T16:50:40Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.4.0/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> and <a href=\"https://rocket.rs/v0.4/news/2018-12-08-version-0.4/\" rel=\"noopener noreferrer\">news article</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.4.1",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.1",
        title: "Rocket v0.4.1",
        date: 2019-05-11T23:56:52Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.4.1/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/v0.4.2",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.2",
        title: "Rocket v0.4.2",
        date: 2019-06-29T01:26:16Z,
        summary: None,
        content: Some(
            "<p>See the <a href=\"https://github.com/SergioBenitez/Rocket/blob/v0.4.2/CHANGELOG.md\" rel=\"noopener noreferrer\">CHANGELOG</a> for more information.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "tag:github.com,2008:Repository/54168759/async-pre-rebase3",
        link: "https://github.com/SergioBenitez/Rocket/releases/tag/async-pre-rebase3",
        title: "async-pre-rebase3",
        date: 2019-12-11T03:17:37Z,
        summary: None,
        content: Some(
            "<p>Revert on_response lifetimes to more closely match 0.4.</p>",
        ),
        media_type: None,
        media_link: None,
        comments_link: None,
    },
]
